pkg/ring/kv/memberlist/memberlist_client.go

```diff
@@ -0,0 +1,788 @@
+			// value changed
+			val, _, err := m.get(key)
+			if err != nil {
+				level.Warn(util.Logger).Log("msg", "failed to decode value while watching for changes", "key", key, "err", err)
+				continue
+			}
+
+			if !f(val) {
+				return
+			}
+
+		case <-m.shutdown:
+			// stop watching on shutdown
+			return
+
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// WatchPrefix watches for any change of values stored under keys with given prefix. When change occurs,
+// function 'f' is called with key and current value.
+// Each change of the key results in one notification. If there are too many pending notifications ('f' is slow),
+// some notifications may be lost.
+//
+// Watching ends when 'f' returns false, context is done, or this client is shut down.
+//
+// Part of kv.Client interface.
+func (m *Client) WatchPrefix(ctx context.Context, prefix string, f func(string, interface{}) bool) {
+	// we use bigger buffer here, since keys are interesting and we don't want to lose them.
+	w := make(chan string, 16)
+
+	// register watcher
+	m.watchersMu.Lock()
+	m.prefixWatchers[prefix] = append(m.prefixWatchers[prefix], w)
+	m.watchersMu.Unlock()
+
+	defer func() {
+		// unregister watcher on exit
+		m.watchersMu.Lock()
+		defer m.watchersMu.Unlock()
+
+		removeWatcherChannel(prefix, w, m.prefixWatchers)
+	}()
+
+	for {
+		select {
+		case key := <-w:
+			val, _, err := m.get(key)
+			if err != nil {
+				level.Warn(util.Logger).Log("msg", "failed to decode value while watching for changes", "key", key, "err", err)
+				continue
+			}
+
+			if !f(key, val) {
+				return
+			}
+
+		case <-m.shutdown:
+			// stop watching on shutdown
+			return
+
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func removeWatcherChannel(k string, w chan string, watchers map[string][]chan string) {
+	ws := watchers[k]
+	for ix, kw := range ws {
+		if kw == w {
+			ws = append(ws[:ix], ws[ix+1:]...)
+			break
+		}
+	}
+
+	if len(ws) > 0 {
+		watchers[k] = ws
+	} else {
+		delete(watchers, k)
+	}
+}
+
+func (m *Client) notifyWatchers(key string) {
+	m.watchersMu.Lock()
+	defer m.watchersMu.Unlock()
+
+	for _, kw := range m.watchers[key] {
+		select {
+		case kw <- key:
+			// notification sent.
+		default:
+			// cannot send notification to this watcher at the moment
+			// but since this is a buffered channel, it means that
+			// there is already a pending notification anyway
+		}
+	}
+
+	for p, ws := range m.prefixWatchers {
+		if strings.HasPrefix(key, p) {
+			for _, pw := range ws {
+				select {
+				case pw <- key:
+					// notification sent.
+				default:
+					c, _ := m.watchPrefixDroppedNotifications.GetMetricWithLabelValues(p)
+					if c != nil {
+						c.Inc()
+					}
+
+					level.Warn(util.Logger).Log("msg", "failed to send notification to prefix watcher", "prefix", p)
+				}
+			}
+		}
+	}
+}
+
+// CAS is part of kv.Client implementation.
+//
+// Unlike other implementation, it doesn't really do Compare-And-Swap. New values returned from function 'f'
```

Comment:
Comment is slightly outdated... this function combines CAS and Merge together. Value is CAS-ed only if Merge returns a change.